### PR TITLE
tabular inference

### PR DESCRIPTION
**Pull request by nutsam**

### CODE REVIEW

File: flavor/serve/inference/data_models/functional/aicoco_data_model.py [hunk 1]
```diff
@@ -58,3 +58,28 @@ class AiCOCOImageFormat(BaseModel, extra="forbid"):
...
     objects: Sequence[AiObject]
     meta: AiMeta
+
+
+class AiTable(BaseModel, extra="allow"):
+    id: str
+    file_name: str
+
+
+class AiInstance(BaseModel):
```
Comment (ding3820) [inline flavor/serve/inference/data_models/functional/aicoco_data_model.py]: `AiInstance(BaseModel, extra="allow")`

File: flavor/serve/inference/data_models/functional/aicoco_data_model.py [hunk 2]
```diff
@@ -58,3 +58,28 @@ class AiCOCOImageFormat(BaseModel, extra="forbid"):
...
+    regressions: Optional[Sequence[AiRegressionItem]]
+
+
+class AiTableMeta(BaseModel, extra="allow"):
+    window_size: int
+
+
+class AiCOCOTabularFormat(BaseModel, extra="forbid"):
+    tables: Sequence[AiTable]
+    categories: Optional[Sequence[AiCategory]]
```
Comment (ding3820) [inline flavor/serve/inference/data_models/functional/aicoco_data_model.py]: `categories`和`regressions`應該不是Optional，至少要是空的list

File: flavor/serve/inference/data_models/functional/aicoco_data_model.py [hunk 3]
```diff
@@ -58,3 +58,28 @@ class AiCOCOImageFormat(BaseModel, extra="forbid"):
...
+class AiTable(BaseModel, extra="allow"):
+    id: str
+    file_name: str
+
+
+class AiInstance(BaseModel):
+    id: str
+    table_id: str
+    row_indexes: Sequence[int]
+    category_ids: Optional[Sequence[AiCategory]]
```
Comment (ding3820) [inline flavor/serve/inference/data_models/functional/aicoco_data_model.py]: Sequence[str]就好

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 4]
```diff
@@ -367,3 +370,154 @@ def __call__(
...
+    def __call__(
+        self, tables: Dict[str, Any], meta: Dict[str, Any], files: Sequence[str], **kwargs
+    ) -> Any:
+        """
+        Run the inference model.
+        """
+        assert len(tables) == len(files), "`files` and `tables` should have same length."
+
+        dataframes = self.data_reader(files, **kwargs)
+        instances = self._set_instances(dataframes, tables, meta)
```
Comment (ding3820) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: 應該要處理files和tables順序不同的問題

File: flavor/serve/inference/data_models/functional/aicoco_data_model.py (hunk 2 above)
Comment (nutsam) [reply flavor/serve/inference/data_models/functional/aicoco_data_model.py]: 這邊之前討論是希望空的時候給 `None`, 而不是 `[]`.
上次討論的結果應該是有寫的都是 `必填`, `Optional` 是可以填 `None`, 但使用者需手動填上去？

File: flavor/serve/inference/data_models/functional/aicoco_data_model.py (hunk 1 above)
Comment (nutsam) [reply flavor/serve/inference/data_models/functional/aicoco_data_model.py]: Done.

File: flavor/serve/inference/data_models/functional/aicoco_data_model.py (hunk 3 above)
Comment (nutsam) [reply flavor/serve/inference/data_models/functional/aicoco_data_model.py]: Done.

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py (hunk 4 above)
Comment (nutsam) [reply flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: sort before read

File: flavor/serve/inference/data_models/functional/aicoco_data_model.py (hunk 2 above)
Comment (nutsam) [reply flavor/serve/inference/data_models/functional/aicoco_data_model.py]: remove optional

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 5]
```diff
@@ -367,3 +370,162 @@ def __call__(
...
+            tables (Sequence[Dict[str, Any]]): List of tables.
+            instances (Sequence[AiInstance]): List of inference instances.
+            meta (Dict[str, Any]): Additional metadata.
+            categories (Optional[Sequence[Dict[str, Any]]]): List of inference categories. Default: None.
+            regressions (Optional[Sequence[Dict[str, Any]]]): List of inference regressions. Default: None.
+
+        Returns:
+            Any: AiCOCO formatted output.
+        """
+        pass
```
Comment (calvin89029) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: raise NotImplementedError

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 6]
```diff
@@ -367,3 +370,162 @@ def __call__(
...
+            num_instances = len(df) // window_size
+            instances = []
+            for i in range(num_instances):
+                instances.append(
+                    {
+                        "table_id": table["id"],
+                        "row_indexes": list(range(i * window_size, (i + 1) * window_size)),
+                    }
+                )
+            table_instances.append(instances)
```
Comment (calvin89029) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: (先留note) decouple check & strategy

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 7]
```diff
@@ -367,3 +370,162 @@ def __call__(
...
+
+        read_func = {
+            ".csv": pd.read_csv,
+            ".parquet": pd.read_parquet,
+            ".xls": pd.read_excel,
+            ".xlsx": pd.read_excel,
+            ".zip": lambda file_name: pd.read_csv(
+                zipfile.ZipFile(file_name, "r").open(zipfile.ZipFile(file_name, "r").namelist()[0])
+            ),
+        }
```
Comment (calvin89029) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: (先留note) 2 issues to discussed
1. do we need to handle reader in flavor
2. extension safe? alternatives?

File: flavor/serve/inference/data_models/functional/__init__.py [hunk 8]
```diff
@@ -2,9 +2,13 @@
     AiAnnotation,
     AiCategory,
     AiCOCOImageFormat,
+    AiCOCOTabularFormat,
     AiImage,
+    AiInstance,
```
Comment (calvin89029) [inline flavor/serve/inference/data_models/functional/__init__.py]: aicoco名詞 看不懂instance, object的含義

File: flavor/serve/inference/strategies/aicoco_strategy.py [hunk 9]
```diff
@@ -688,3 +699,238 @@ def update_images_meta(
...
+        categories: Optional[Sequence[Dict[str, Any]]] = None,
+        regressions: Optional[Sequence[Dict[str, Any]]] = None,
+    ) -> AiCOCOTabularOut:
+        """
+        Prepare prerequisite for AiCOCO.
+
+        Args:
+            tables (Sequence[Dict]): List of AiCOCO table compatible dict.
+            instances (Sequence[Dict]): List of AiCOCO instance compatible dict.
+            meta (Optional[Dict[str, Any]]): AiCOCO tabular meta dict.
```
Comment (calvin89029) [inline flavor/serve/inference/strategies/aicoco_strategy.py]: not Optional?

File: pyproject.toml [hunk 10]
```diff
@@ -18,6 +18,7 @@ gradio = "^4.0.0"
 requests = "^2.28.0"
 opencv-python-headless = "^4.5.3.56"
 tritonclient = {version = "^2.17.0", extras = ["http"]}
+pandas = "^2.2.2"
```
Comment (calvin89029) [inline pyproject.toml:22]: example review借留：補example還有tabular相關的spec

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py (hunk 6 above)
Comment (nutsam) [reply flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: remove `set_instance`, generate instance from strategy.

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py (hunk 7 above)
Comment (nutsam) [reply flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: remove `data_reader` from flavor

File: flavor/serve/inference/strategies/aicoco_strategy.py (hunk 9 above)
Comment (nutsam) [reply flavor/serve/inference/strategies/aicoco_strategy.py]: remove Optional from docstring

File: examples/inference/README.md [hunk 11]
```diff
@@ -141,10 +141,14 @@ We try to integrate FLaVor with other existing projects or libraries. Please vis
...
 - [3D Segmentation task example](./seg3d_example.ipynb)
+- [Tabular classification task example](./tabular_cls_example.ipynb)
+- [Tabular regression task example](./tabular_reg_example.ipynb)
 
 ## Other Things You Might Want to Know
 
 - [AiCOCO format specification](./docs/AiCOCO_spec.md)
+- [AiCOCO tabular format specification](./docs/AiCOCO_tabular_spec.md)
 - [Standard input and output structure](./docs/input_output_structure.md)
+- [Standard input and output structure (tabular)](./docs/input_output_structure(tabular).md)
```
Comment (calvin89029) [inline examples/inference/README.md]: 改aicoco "image" sp...
Standard input and output structure (image)
區分清楚一點

File: examples/inference/README.md [hunk 12]
```diff
@@ -141,10 +141,14 @@ We try to integrate FLaVor with other existing projects or libraries. Please vis
 - [Regression task example](./reg_example.ipynb)
 - [Segmentation task example](./seg_example.ipynb)
 - [3D Segmentation task example](./seg3d_example.ipynb)
+- [Tabular classification task example](./tabular_cls_example.ipynb)
+- [Tabular regression task example](./tabular_reg_example.ipynb)
```
Comment (calvin89029) [inline examples/inference/README.md:145]: 先note，不確定是沒push到檔案還是還沒好先寫readme

File: examples/inference/docs/AiCOCO_tabular_spec.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+      "id": "nanoid(21)",
+      "name": "SBP",
+      "superregression_id": "nanoid(21)" or null,
+      // Optional fields
+      "unit": "mmHg",
+      "threshold": "140",
+      ...
+    }
+  ],
+  "instances": [
```
Comment (calvin89029) [inline examples/inference/docs/AiCOCO_tabular_spec.md]: cntl-f 總共好像有13處沒改到

File: examples/inference/README.md (hunk 11 above)
Comment (nutsam) [reply examples/inference/README.md]: done.

File: examples/inference/README.md (hunk 12 above)
Comment (nutsam) [reply examples/inference/README.md:145]: sorry, `.gitignore` 擋到 `*.ipynb`, 補上去了

File: examples/inference/docs/AiCOCO_tabular_spec.md (hunk 13 above)
Comment (nutsam) [reply examples/inference/docs/AiCOCO_tabular_spec.md]: done.

File: pyproject.toml (hunk 10 above)
Comment (nutsam) [reply pyproject.toml:22]: done.

File: pyproject.toml [hunk 14]
```diff
@@ -88,6 +89,12 @@ optional = true
 pre-commit = "^3.5.0"
```
Comment (calvin89029) [inline pyproject.toml:94]: 借留 test幫忙補一下

File: examples/inference/README.md [hunk 15]
```diff
@@ -136,15 +136,19 @@ Once the FLaVor Inference Service is running, send POST requests to the `/invoca
 
 We try to integrate FLaVor with other existing projects or libraries. Please visit following instruction pages:
 
-- [Classification task example](./cls_example.ipynb)
+- [Image Classification task example](./cls_example.ipynb)
 - [Detection task example](./det_example.ipynb)
```
Comment (ding3820) [inline examples/inference/README.md]: 其他task也幫加Image 感謝

File: examples/inference/README.md [hunk 16]
```diff
@@ -136,15 +136,19 @@ Once the FLaVor Inference Service is running, send POST requests to the `/invoca
...
+- [Tabular classification task example](./tabular_cls_example.ipynb)
+- [Tabular regression task example](./tabular_reg_example.ipynb)
 
 ## Other Things You Might Want to Know
 
-- [AiCOCO format specification](./docs/AiCOCO_spec.md)
-- [Standard input and output structure](./docs/input_output_structure.md)
+- [AiCOCO image format specification](./docs/AiCOCO_spec.md)
+- [AiCOCO tabular format specification](./docs/AiCOCO_tabular_spec.md)
+- [Standard input and output structure (image)](./docs/input_output_structure.md)
```
Comment (ding3820) [inline examples/inference/README.md]: 檔名沒改到

File: examples/inference/docs/input_output_structure(tabular).md [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+        categories: Sequence[Dict[str, Any]],
+        meta: Dict[str, Any],
+        **kwargs,
+    ) -> AiCOCOTabularOut:
+```
+
+### Parameters
+- `model_out`: np.ndarray
+
+   1D NumPy array with classification results, the output shape will be :
```
Comment (ding3820) [inline examples/inference/docs/input_output_structure(tabular).md]: 2D NumPy array with classification results in shape of (n, c), where `n` represents n instances and `c` is the number of classes. Here are some output examples:

File: examples/inference/docs/input_output_structure(tabular).md [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+        regressions: Sequence[Dict[str, Any]],
+        meta: Dict[str, Any],
+        **kwargs,
+    ) -> AiCOCOTabularOut:
+```
+
+### Parameters
+- `model_out`: np.ndarray
+
+   1D NumPy array with regression results
```
Comment (ding3820) [inline examples/inference/docs/input_output_structure(tabular).md]: Modify as above classification example

File: examples/inference/README.md (hunk 15 above)
Comment (nutsam) [reply examples/inference/README.md]: done.

File: examples/inference/README.md (hunk 16 above)
Comment (nutsam) [reply examples/inference/README.md]: done.

File: examples/inference/docs/input_output_structure(tabular).md (hunk 17 above)
Comment (nutsam) [reply examples/inference/docs/input_output_structure(tabular).md]: done.

File: examples/inference/docs/input_output_structure(tabular).md (hunk 18 above)
Comment (nutsam) [reply examples/inference/docs/input_output_structure(tabular).md]: done.

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 19]
```diff
@@ -367,3 +368,122 @@ def __call__(
...
+            records (Sequence[AiRecord]): List of inference records.
+            meta (Dict[str, Any]): Additional metadata.
+            categories (Optional[Sequence[Dict[str, Any]]]): List of inference categories. Default: None.
+            regressions (Optional[Sequence[Dict[str, Any]]]): List of inference regressions. Default: None.
+
+        Returns:
+            Any: AiCOCO formatted output.
+        """
+        raise NotImplementedError
+
```
Comment (ding3820) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py:518]: inference和postprocess不寫嗎

File: flavor/serve/inference/strategies/aicoco_strategy.py [hunk 20]
```diff
@@ -688,3 +717,245 @@ def update_images_meta(
...
+
+        tables = [AiTable(**table) for table in tables]
+        categories = categories if categories is not None else []
+        regressions = regressions if regressions is not None else []
+
+        self.aicoco_records = self.generate_records(dataframes, tables, meta)
+
+        if not hasattr(self, "aicoco_categories") and not hasattr(self, "aicoco_regressions"):
+            # only activate at first run
+            self.aicoco_categories = self.generate_categories(categories)
```
Comment (ding3820) [inline flavor/serve/inference/strategies/aicoco_strategy.py:794]: 這邊是不是就不用掛`self`了

Comment (ding3820) [reply flavor/serve/inference/strategies/aicoco_strategy.py:794]: 上面也幫改 感謝

File: tests/inference/test_tasks/tabular_reg_example.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        return None
+
+    def set_regressions(self) -> None:
+        regressions = [{"name": "reg_value"}]
+        return regressions
+
+    def data_reader(
+        self, tables: Dict[str, Any], files: Sequence[str], **kwargs
+    ) -> List[pd.DataFrame]:
+        table_names = [table["file_name"].replace("/", "_") for table in tables]
```
Comment (ding3820) [inline tests/inference/test_tasks/tabular_reg_example.py:43]: L43-47移出到`__call__`裡的`data_reader`前

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py (hunk 19 above)
Comment (nutsam) [reply flavor/serve/inference/inference_models/base_aicoco_inference_model.py:518]: 已補.

File: flavor/serve/inference/strategies/aicoco_strategy.py (hunk 20 above)
Comment (nutsam) [reply flavor/serve/inference/strategies/aicoco_strategy.py:794]: 要掛, only activate at first run, 才不用每次 `__call__` 都跑

File: tests/inference/test_tasks/tabular_reg_example.py (hunk 21 above)
Comment (nutsam) [reply tests/inference/test_tasks/tabular_reg_example.py:43]: done.

File: flavor/serve/inference/strategies/aicoco_strategy.py (hunk 20 above)
Comment (ding3820) [reply flavor/serve/inference/strategies/aicoco_strategy.py:794]: ＯＫ 忘了

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 22]
```diff
@@ -367,3 +368,179 @@ def __call__(
...
+            sorted_files (Sequence[str]): sorted files.
+
+        """
+        table_names = [table["file_name"].replace("/", "_") for table in tables]
+
+        table_dict = {table["file_name"].replace("/", "_"): table for table in tables}
+        file_dict = {file: file for file in files}
+
+        table_names = sorted(table_names, key=lambda s: s[::-1])
+        file_names = sorted(files, key=lambda s: s[::-1])
```
Comment (ding3820) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: 可簡化成兩行sorted

Comment (nutsam) [reply flavor/serve/inference/inference_models/base_aicoco_inference_model.py]: done.

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 23]
```diff
@@ -367,3 +368,172 @@ def __call__(
...
+            files (Sequence[str]): List of input file_names.
+
+        Returns:
+            sorted_tables (Dict[str, Any]): sorted tables.
+            sorted_files (Sequence[str]): sorted files.
+
+        """
+
+        sorted_tables = sorted(tables, key=lambda x: x["file_name"].replace("/", "_")[::-1])
+        sorted_files = sorted(files, key=lambda x: x[::-1])
```
Comment (calvin89029) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py:402]: 需檢查files是否和table一致

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 24]
```diff
@@ -367,3 +368,172 @@ def __call__(
...
+
+        return self.network(x)
+
+    def postprocess(self, out: Any) -> Any:
+        """
+        A default operation for post-processing which is identical transformation.
+
+        Override it if you need activations like softmax or sigmoid generating the prediction.
+
+        Args:
```
Comment (calvin89029) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py:482]: preprocess, inference, postprocess在BaseAiCOCOTabularInferenceModel, BaseAiCOCOImagenferenceModel看起來一模一樣，為何不寫至BaseAiCOCOInferenceModel？BaseAiCOCOInferenceModel寫nonimplementerror，除非是其他AiCOCO model會來override然後有不同預設選擇

File: pyproject.toml [hunk 25]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "flavor"
-version = "2.0.0"
+version = "2.0.1"
```
Comment (calvin89029) [inline pyproject.toml]: 這不是小bugfix 打 2.1.0

**Comment by calvin89029** [comment]
最後如果 @ding3820 也review完成，麻煩 @nutsam 補上test跑過的截圖，thx

**Comment by nutsam** [comment]
補測試成功圖

截圖 2024-08-08 上午10 49 32

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py (hunk 23 above)
Comment (nutsam) [reply flavor/serve/inference/inference_models/base_aicoco_inference_model.py:402]: done.

File: pyproject.toml (hunk 25 above)
Comment (nutsam) [reply pyproject.toml]: done.

File: flavor/serve/inference/inference_models/base_aicoco_inference_model.py [hunk 26]
```diff
@@ -367,3 +368,177 @@ def __call__(
...
+        window_size = meta["window_size"]
+        assert all(
+            len(df) % window_size == 0 for df in dataframes
+        ), f"Not all DataFrames have a length that is divisible by {window_size}"
+
+    def data_reader(
+        self, tables: Dict[str, Any], files: Sequence[str], **kwargs
+    ) -> Sequence[pd.DataFrame]:
+        """
+        Read data for inference model.
```
Comment (ding3820) [inline flavor/serve/inference/inference_models/base_aicoco_inference_model.py:434]: 再幫補個return value的詳細敘述，可參考https://github.com/ailabstw/FLaVor/blob/dev/infer-tabular/flavor/serve/inference/inference_models/base_aicoco_inference_model.py#L258

**Review by calvin89029 — APPROVED** [review]

**Pull request merged**